include/micm/jit/jit_function.hpp

```diff
@@ -0,0 +1,307 @@
+    JitLoop StartLoop(std::string name, llvm::Value *start, llvm::Value *end, llvm::Value *step);
+
+    /// @brief End a loop block
+    /// @param loop Loop reference
+    void EndLoop(JitLoop& loop);
+
+   private:
+    llvm::AllocaInst* CreateEntryBlockAlloca(llvm::Type* type, const std::string& var_name);
+  };
+
+  class JitFunctionBuilder
+  {
+    std::shared_ptr<JitCompiler> compiler_;
+    std::string name_;
+    std::vector<std::pair<std::string, JitType>> arguments_;
+    JitType return_type_{ JitType::Void };
+    friend class JitFunction;
+
+   public:
+    JitFunctionBuilder() = delete;
+    JitFunctionBuilder(std::shared_ptr<JitCompiler> compiler);
+    JitFunctionBuilder& name(std::string name);
+    JitFunctionBuilder& arguments(const std::vector<std::pair<std::string, JitType>>& arguments);
+    JitFunctionBuilder& return_type(JitType type);
+  };
+
+  inline JitFunctionBuilder JitFunction::create(std::shared_ptr<JitCompiler> compiler)
+  {
+    return JitFunctionBuilder{ compiler };
+  }
+
+  JitFunction::JitFunction(JitFunctionBuilder& function_builder)
+      : name_(function_builder.name_),
+        compiler_(function_builder.compiler_),
+        context_(std::make_unique<llvm::LLVMContext>()),
+        module_(std::make_unique<llvm::Module>(name_ + " module", *context_)),
+        builder_(std::make_unique<llvm::IRBuilder<>>(*context_)),
+        exit_on_error_(),
+        arguments_(),
+        function_(),
+        entry_block_()
+  {
+    module_->setDataLayout(compiler_->GetDataLayout());
+
+    // Prototype function
+    std::vector<llvm::Type*> arg_types;
+    for (const auto& pair : function_builder.arguments_)
+    {
+      llvm::Type* type = GetType(pair.second);
+      arg_types.push_back(type);
+      arguments_.push_back({ .name_ = pair.first, .type_ = type });
+    }
+    llvm::FunctionType* function_type = llvm::FunctionType::get(GetType(function_builder.return_type_), arg_types, false);
+    function_ = llvm::Function::Create(function_type, llvm::Function::ExternalLinkage, name_, module_.get());
+    llvm::Function::arg_iterator arg_iter = function_->arg_begin();
+    for (auto& arg : arguments_)
+    {
+      arg.arg_ = arg_iter++;
+      arg.arg_->setName(arg.name_);
+    }
+
+    // function body
+
+    // set up entry block
+    entry_block_ = llvm::BasicBlock::Create(*context_, "entry", function_);
+    builder_->SetInsertPoint(entry_block_);
+
+    // set up function argument variables
+    for (auto& arg : arguments_)
+      arg.alloca_ = CreateEntryBlockAlloca(arg.type_, arg.name_);
+    for (auto& arg : arguments_)
+      builder_->CreateStore(arg.arg_, arg.alloca_);
+    for (auto& arg : arguments_)
+      arg.ptr_ = builder_->CreateLoad(arg.type_, arg.alloca_);
+  }
+
+  std::pair<llvm::orc::ResourceTrackerSP, llvm::JITTargetAddress> JitFunction::Generate()
+  {
+    std::pair<llvm::orc::ResourceTrackerSP, llvm::JITTargetAddress> ret_val;
```

Comment:
can we add an assert or some other check to ensure this is only called once?

Reply:
done!